File: oldnewthing/oldnewthing.go
package oldnewthing

import (
	"fmt"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/gorilla/feeds"

	"github.com/myhro/feeds/errormap"
	"github.com/myhro/feeds/generator"
)

const Command = "oldnewthing"
const FeedTitle = "The Old New Thing"

func CleanDescription(s *goquery.Selection) string {
	s = s.Clone()
	s.Find(".entry-header").Remove()
	s.Find(".entry-footer").Remove()

	return strings.TrimSpace(s.Text())
}

func XML() (string, error) {
	gen := generator.Generator{
		CSS:   ".entry-area",
		Title: FeedTitle,
		URL:   "https://devblogs.microsoft.com/oldnewthing/",
	}

	gen.Parse = func(i int, s *goquery.Selection) {
		if i >= 10 {
			return
		}

		date := s.Find(".entry-post-date").Text()
		date = strings.TrimSpace(date)
		title := s.Find(".entry-title").Text()
		link, _ := s.Find(".entry-title").Find("a").Attr("href")
		description := CleanDescription(s.Find(".entry-content"))

		created, err := time.Parse("January 2, 2006", date)
		if err != nil {
			errormap.Store(Command, fmt.Errorf("time.Parse: %w", err))
		}

		item := &feeds.Item{
			Title:       title,
			Created:     created,
			Link:        &feeds.Link{Href: link},
			Description: description,
		}
		gen.Feed.Items = append(gen.Feed.Items, item)
	}

	atom, err := gen.Generate()
	if err != nil {
		return "", fmt.Errorf("gen.Generate: %w", err)
	}

	return atom, nil
}

File: bin/oldnewthing.ts
import fetch from 'node-fetch';
import { load } from 'cheerio';
import { Feed } from 'feed';

(async () => {
  const url = 'https://devblogs.microsoft.com/oldnewthing/';
  const feed = new Feed({
    id: url,
    title: 'The Old New Thing',
    copyright: '',
    link: url,
  });

  const data = await fetch(url);
  const $ = load(await data.text());
  const posts = $('.masonry-card').slice(0, 10);
  for (const p of posts) {
    const title = $('h3', p).text().trim();
    const link = $('h3 a', p).attr('href') || '';
    const dateElem = $('.justify-content-between > div:first-child', p);
    const date = new Date(dateElem.text().trim());
    const content = $('p.mb-24', p).html() || '';
    const description = content.trim();

    feed.addItem({
      title: title,
      date: date,
      link: link,
      description: description,
    });
  }

  const atom = feed.atom1();
  console.log(atom);
})();
